src/core/streams/stream_actions.js

```diff
@@ -30,5 +30,14 @@ export const StreamActions = {
       targetElement.innerHTML = ""
       targetElement.append(this.templateContent)
     })
+  },
+
+  refresh() {
+    const requestId = this.getAttribute("request-id")
+    const isRecentRequest = requestId && window.Turbo.recentRequests.has(requestId)
+    if (!isRecentRequest) {
+      window.Turbo.cache.exemptPageFromPreview()
+      window.Turbo.visit(window.location.href, { action: "replace" })
```

Comment:
Importing would be nicer, but unfortunately it creates a circular dependency:

```
(!) Circular dependency
src/core/index.js -> src/core/streams/stream_actions.js -> src/core/index.js
created dist/turbo.es2017-umd.js, dist/turbo.es2017-esm.js in 225ms
```

Reply:
Ah, thanks for exploring that @afcapel.

Since `Turbo` isn't really an instance of a class or namespace, but rather a collection of other exports from the `export * as Turbo` statement, what do you think about importing the individual components themselves?

I've had some success with this diff locally:

```diff
diff --git a/src/core/index.js b/src/core/index.js
index 743380d..34269c5 100644
--- a/src/core/index.js
+++ b/src/core/index.js
@@ -12,8 +12,6 @@ const recentRequests = new LimitedSet(20)
 const { navigator } = session
 export { navigator, session, cache, recentRequests, PageRenderer, PageSnapshot, FrameRenderer }
 
-export { StreamActions } from "./streams/stream_actions"
-
 /**
  * Starts the main session.
  * This initialises any necessary observers such as those to monitor
diff --git a/src/core/streams/stream_actions.js b/src/core/streams/stream_actions.js
index 62801fd..e8e20bd 100644
--- a/src/core/streams/stream_actions.js
+++ b/src/core/streams/stream_actions.js
@@ -1,3 +1,5 @@
+import { cache, recentRequests, visit } from "../"
+
 export const StreamActions = {
   after() {
     this.targetElements.forEach((e) => e.parentElement?.insertBefore(this.templateContent, e.nextSibling))
@@ -34,10 +36,10 @@ export const StreamActions = {
 
   refresh() {
     const requestId = this.getAttribute("request-id")
-    const isRecentRequest = requestId && window.Turbo.recentRequests.has(requestId)
+    const isRecentRequest = requestId && recentRequests.has(requestId)
     if (!isRecentRequest) {
-      window.Turbo.cache.exemptPageFromPreview()
-      window.Turbo.visit(window.location.href, { action: "replace" })
+      cache.exemptPageFromPreview()
+      visit(window.location.href, { action: "replace" })
     }
   }
 }
diff --git a/src/index.js b/src/index.js
index b76e238..336101a 100644
--- a/src/index.js
+++ b/src/index.js
@@ -8,6 +8,7 @@ import * as Turbo from "./core"
 window.Turbo = Turbo
 Turbo.start()
 
+export { StreamActions } from "./core/streams/stream_actions"
 export * from "./core"
 export * from "./elements"
 export * from "./http"
```

According to the [official documentation](https://turbo.hotwired.dev/handbook/streams#custom-actions), applications that need to extend `StreamActions` should be importing it directly from the module instead of accessing it through `window.Turbo.StreamActions`.

Reply:
As an alternative to (or in addition to!) resolving the circular dependency and exports, have you considered extending the shape of the `visit` function's optional `VisitOption` interface to accept a `requestId`? Passing that option along could help to encapsulate these details and limit the amount of modules that are dependent on the `cache` and `recentRequests` variables. That might look something like:

```js
import { visit } from "../"

// ...
refresh() {
  visit(this.baseURI, {
    action: "replace",
    requestId: this.getAttribute("request-id")
  })
}
```

Then, the `visit(url, options)` implementation could check for the presence of `options.requestId` to cross reference the `recentRequests`, then call `cache.exemptPageFromPreview()` if necessary.

Reply:
I've opened https://github.com/hotwired/turbo/pull/1026 to explore an alternative to keeping this logic in `StreamActions`. It leans more heavily on encapsulating the logic within the `Session` class. It avoids the [circular dependency trap](https://github.com/hotwired/turbo/pull/1019/files#r1349736505) described above.